Alc/effects/autowah.c

```diff
@@ -121,32 +135,37 @@ static ALvoid ALautowahState_update(ALautowahState *state, const ALCcontext *con
     state->FreqMinNorm   = MIN_FREQ/device->Frequency;
     state->BandwidthNorm = (MAX_FREQ - MIN_FREQ)/device->Frequency;
 
-    CalcAngleCoeffs(0.0f, 0.0f, 0.0f, coeffs);
-    ComputeDryPanGains(&device->Dry, coeffs, slot->Params.Gain, state->TargetGains);
+    STATIC_CAST(ALeffectState,state)->OutBuffer = device->FOAOut.Buffer;
+    STATIC_CAST(ALeffectState,state)->OutChannels = device->FOAOut.NumChannels;
+    for(i = 0;i < MAX_EFFECT_CHANNELS;i++)
+        ComputeFirstOrderGains(&device->FOAOut, IdentityMatrixf.m[i],
+                               slot->Params.Gain, state->Chans[i].TargetGains);
 }
 
 static ALvoid ALautowahState_process(ALautowahState *state, ALsizei SamplesToDo, const ALfloat (*restrict SamplesIn)[BUFFERSIZE], ALfloat (*restrict SamplesOut)[BUFFERSIZE], ALsizei NumChannels)
 {
-    ALfloat *restrict BufferOut = state->BufferOut;
-    ALsizei i;
+    ALfloat (*restrict BufferOut)[BUFFERSIZE] = state->BufferOut;
+    ALsizei c, i;
 
-    for(i = 0;i < SamplesToDo;i++)
+    for(c = 0;c < CHANNELS; c++)
     {
-        ALfloat env_out, f0norm, temp;
-
-        env_out = envelope_follower(state, SamplesIn[0][i]);
-        f0norm = state->BandwidthNorm*env_out + state->FreqMinNorm;
-
-        BiquadFilter_setParams(&state->Filter, BiquadType_Peaking, state->ResonanceGain,
-                               f0norm, 1.0f/Q_FACTOR);
-        BiquadFilter_process(&state->Filter, &temp, &SamplesIn[0][i], 1);
-
-        BufferOut[i] = temp;
+        for(i = 0;i < SamplesToDo;i++)
+        {
+            ALfloat env_out, f0norm, temp;
+
+            env_out = envelope_follower(state, SamplesIn[c][i], c);
+            f0norm = state->BandwidthNorm*env_out + state->FreqMinNorm;
+
+            BiquadFilter_setParams(&state->Chans[c].Filter, BiquadType_Peaking,
+                                   state->ResonanceGain, f0norm, 1.0f/Q_FACTOR);
+            BiquadFilter_process(&state->Chans[c].Filter, &temp, &SamplesIn[c][i], 1);
```

Comment:
>You can generate one envelope from that channel, using it to apply equal filters to all the channels.

Envelope followers always depends of the signal gain, we should correct the gains for the other channels with the appropiate factor if we only use the first channel envelope. The current variations are due to the different gains of each channel, producing a similar effect to the sound of wind in the trees.

Reply:
> Envelope followers always depends of the signal gain, we should correct the gains for the other channels with the appropiate factor if we only use the first channel envelope.

The overall signal gain is stored in the first channel, and the remaining channels control the direction. Changing the other channels won't change the overall signal level, just the direction it's focused in.

If what you're hoping for is different directions to have different filter strengths simultaneously (e.g. a quiet sound on the left having a more closed filter, at the same time as a loud sound on the right having a more open filter), that will need more work. You could convert the B-Format input to A-Format, process the discrete A-Format channels, then convert back to B-Format, similar to how reverb works. Though that may not sound very good since that's essentially trying to unmix the individual sounds and apply different filters to them. The spatial resolution isn't very high to clearly distinguish nearby sounds from each other (to say nothing of sounds that overlap).

If that's what the app is wanting, it would be better for it to create multiple autowah effects, pass each source through a different one so each is affected uniquely, then let them mix in the output. With each effect retaining 3D positioning with 3D processing, the sound directions remain intact.

Reply:
> Luckily, B-Format also offers precisely what's needed to uniformly affect the whole output. The first channel (index 0) has the omnidirectional response, which is the total amplitude of the signal reaching the listener, irrespective of direction. You can generate one envelope from that channel, using it to apply equal filters to all the channels. So long as each channel uses the same filter parameters for a given sample time, the phase alignment remains intact (all channels will see the same relative phase offset for a given frequency).

The solution that you propose here can work well, I like it.

Another solution could be to process channel 0 and leave the directional channels unprocessed (without going through to the autowah filter), so that we would have 3D effect, but with the classic autowah.

I left the 3D processing disabled because it wasn't clear which method was better and I preferred to comment here with you. I think that the musicians and guitarists who use these effects (modulator, distortion, autowah,...) using the Creative hardware will want a more classic effect and not get into 3D processing, this autowah implementation with 3D processing may have a plus that can be interesting to them.

When I was doing this I thought whether would be good to have control over the activation/deactivation of 3D processing in the EFX extension (for all effects), via configuration file. Something like:

`3D_EFX = False `

In alsoft. ini.

Reply:
> Another solution could be to process channel 0 and leave the directional channels unprocessed (without going through to the autowah filter), so that we would have 3D effect, but with the classic autowah.

That unfortunately won't work since it will still cause phase misalignment; the filter will still apply phase shifts to channel 0 and not the other channels. It will also cause amplitude to misalign (directionality is determined by the directional components' values relative to the omnidirectional component's values). Even if phase wasn't a problem, increasing the amplitude of certain frequencies in the first channel while leaving the other channels alone would cause the affected frequencies to lose directionality.

> I left the 3D processing disabled because it wasn't clear which method was better and I preferred to comment here with you.

Personally, I'd prefer either that last patch for 3D processing be removed, or see it redone (generating the envelope from channel 0 only, then applying the same filter parameters to each channel). It's not worth it to have a compile-time toggle in this case, either. A configuration option can be looked at later if it's really desirable.

It may also be a good idea to rebase your branch against master, to get rid of the extraneous merge commit.

Reply:
>Personally, I'd prefer either that last patch for 3D processing be removed, or see it redone (generating the envelope from channel 0 only, then applying the same filter parameters to each channel). It's not worth it to have a compile-time toggle in this case, either. A configuration option can be looked at later if it's really desirable.

This option works very well, I tested it and seems doesn't has problems in the processed sound :) .

>It may also be a good idea to rebase your branch against master, to get rid of the extraneous merge commit.

The f****** Github desktop generates those "extraneous" commits, and don't have an option for rebase. I don't have much experience on git using command prompt, so I can't delete them. How can it be done?

Reply:
I might be able to make github rebase the commits automatically when doing the merge. I normally just click the merge button, so let's see what happens.